File: sdks/js/src/sprites-compat.ts
/**
 * `@pilots/sdk/sprites-compat` -- the sprites-shaped face of the pilots API.
 *
 * It exists so the reference customer moves providers by changing one import
 * line. The surface is exactly what a sprites client consumes and nothing
 * more: no sessions, no `/control` multiplex, no filesystem API.
 *
 * Two rules decide the shapes here.
 *
 *   - A sprite's `id` is the machine's NAME. A sprites consumer persists the
 *     id and then hands it back as a path segment, and the alias that serves
 *     those paths resolves names. `machineId` carries the `m-...` id for
 *     anyone who wants the raw client.
 *   - A restore is IN PLACE. `restoreCheckpoint` performs exactly one request,
 *     `POST /v1/checkpoints/{id}/restore`, and never creates a machine: a
 *     machine created in a restore would get a new URL, and a URL is permanent.
 */

import { PilotsClient } from './client.ts'
import { NotFoundError, PilotsError } from './errors.ts'
import { resolveBaseURL } from './http.ts'
import type { ExecStream, ExecStreamOptions, WebSocketCtor } from './stream.ts'
import type { Machine } from './types.ts'

/** What a sprites consumer destructures from an exec. */
export interface ExecResult {
  stdout: string
  stderr: string
  exitCode: number
}

export interface SpriteInfo {
  id: string
  name: string
  url: string
  status: string
}

export interface Checkpoint {
  id: string
  createTime: Date
  comment?: string
}

export interface ExecOptions {
  cwd?: string
  env?: Record<string, string>
  /** Milliseconds, as sprites spells it. Sent as `timeout_ms`. */
  timeout?: number
}

export interface SpritesClientOptions {
  baseURL?: string
  /** Milliseconds. Default 30000. */
  timeout?: number
  WebSocket?: WebSocketCtor
}

const DEFAULT_TIMEOUT = 30_000
/** A machine id as the manager mints them. */
const MACHINE_ID = /^m-[0-9a-f]{24}$/

export class SpritesClient {
  readonly baseURL: string
  readonly token: string
  readonly timeout: number
  /** The underlying typed client, for anything this surface does not cover. */
  readonly pilots: PilotsClient

  private readonly byName = new Map<string, Machine>()

  constructor(token: string, opts: SpritesClientOptions = {}) {
    this.token = token
    this.baseURL = resolveBaseURL(opts.baseURL)
    this.timeout = opts.timeout ?? DEFAULT_TIMEOUT
    this.pilots = new PilotsClient(token, {
      baseURL: this.baseURL,
      timeoutMs: this.timeout,
      ...(opts.WebSocket ? { WebSocket: opts.WebSocket } : {}),
    })
  }

  /** A handle. No network until something is asked of it. */
  sprite(name: string): Sprite {
    return new Sprite(name, this)
  }

  /**
   * Creates a machine. The body is `{name}` and nothing else: the router
   * already proxies every machine's URL to the guest's port 8080, which is
   * the port a dev server binds, so there is no port to declare.
   */
  async createSprite(name: string): Promise<Sprite> {
    const machine = await this.pilots.machines.create({ name })
    this.byName.set(machine.name, machine)
    return new Sprite(machine.name, this, machine)
  }

  async getSprite(name: string): Promise<Sprite> {
    const machine = await this.resolve(name)
    return new Sprite(machine.name, this, machine)
  }

  async deleteSprite(name: string): Promise<void> {
    const machine = await this.resolve(name)
    this.byName.delete(machine.name)
    await this.pilots.machines.destroy(machine.id)
  }

  /**
   * A no-op. A workload's URL is public by default here, so there is nothing
   * to switch on; it stays on the surface because a sprites consumer calls it.
   */
  async setPublicUrl(_name: string): Promise<void> {
    return
  }

  /** Drops a cached machine, so the next resolve goes back to the fleet. */
  forget(name: string): void {
    this.byName.delete(name)
  }

  /**
   * Name to machine. Names are fleet-unique, so a list filtered on the name
   * is the lookup; an argument that matches no name but looks like a machine
   * id is tried as one, which is what makes an id work wherever a name does.
   */
  async resolve(name: string): Promise<Machine> {
    const cached = this.byName.get(name)
    if (cached) return cached

    const machines = await this.pilots.machines.list()
    for (const machine of machines) this.byName.set(machine.name, machine)
    const found = this.byName.get(name)
    if (found) return found

    if (MACHINE_ID.test(name)) {
      const machine = await this.pilots.machines.get(name)
      this.byName.set(machine.name, machine)
      return machine
    }
    throw new NotFoundError(`no sprite named ${name}`)
  }
}

export class Sprite {
  readonly name: string

  private readonly client: SpritesClient
  private machine: Machine | undefined

  constructor(name: string, client: SpritesClient, machine?: Machine) {
    this.name = name
    this.client = client
    this.machine = machine
  }

  /** The machine's NAME, which is what a sprites consumer persists. */
  get id(): string {
    return this.name
  }

  /** `Machine.url` verbatim. Permanent: it survives every lifecycle step. */
  get url(): string {
    return this.machine?.url ?? ''
  }

  get status(): string {
    return this.machine?.state ?? ''
  }

  /** `m-<24 hex>`, for calls made through the raw `PilotsClient`. */
  get machineId(): string {
    return this.machine?.id ?? ''
  }

  toInfo(): SpriteInfo {
    return { id: this.id, name: this.name, url: this.url, status: this.status }
  }

  /** Runs `[file, ...args]`, each argument shell-quoted, under the guest's bash. */
  execFile(file: string, args: string[] = [], opts: ExecOptions = {}): Promise<ExecResult> {
    return this.exec(shellQuote([file, ...args]), opts)
  }

  async exec(command: string, opts: ExecOptions = {}): Promise<ExecResult> {
    const res = await this.withMachine((machine) =>
      this.client.pilots.machines.exec(machine.id, {
        cmd: command,
        ...(opts.cwd !== undefined ? { cwd: opts.cwd } : {}),
        ...(opts.env !== undefined ? { env: opts.env } : {}),
        ...(opts.timeout !== undefined ? { timeout_ms: opts.timeout } : {}),
      }),
    )
    return { stdout: res.stdout, stderr: res.stderr, exitCode: res.exit_code }
  }

  /**
   * Streams a command. `stdin` stays off unless asked for.
   *
   * Synchronous, as sprites' is, which means the machine has to be known
   * already: use `createSprite` or `getSprite` rather than the lazy `sprite`.
   */
  spawn(command: string, args: string[] = [], opts: ExecStreamOptions = {}): ExecStream {
    if (!this.machine) {
      throw new PilotsError(
        `spawn needs a resolved sprite: await client.getSprite('${this.name}') first`,
      )
    }
    return this.client.pilots.machines.execStream(this.machine.id, [command, ...args], opts)
  }

  /**
   * Returns a `Response` whose body is one NDJSON line, the checkpoint's JSON.
   * A sprites consumer reads `.text()` and scans the lines backwards for `id`.
   */
  async createCheckpoint(comment?: string): Promise<Response> {
    const checkpoint = await this.withMachine((machine) =>
      this.client.pilots.machines.checkpoint(machine.id, comment !== undefined ? { comment } : {}),
    )
    return ndjsonResponse(checkpoint)
  }

  async listCheckpoints(): Promise<Checkpoint[]> {
    const list = await this.withMachine((machine) =>
      this.client.pilots.machines.listCheckpoints(machine.id),
    )
    return list.map((c) => ({
      id: c.id,
      createTime: new Date(c.created_at * 1000),
      ...(c.comment !== undefined ? { comment: c.comment } : {}),
    }))
  }

  /**
   * Restores IN PLACE. Exactly one request, and no machine is created: the
   * same machine keeps its id, its URL and its agent token.
   */
  async restoreCheckpoint(id: string): Promise<Response> {
    const machine = await this.client.pilots.checkpoints.restore(id)
    this.machine = machine
    return ndjsonResponse(machine)
  }

  destroy(): Promise<void> {
    return this.client.deleteSprite(this.name)
  }

  /**
   * Resolves the machine, runs the call, and on a 404 forgets what it knew and
   * tries once more -- a cached machine that was destroyed and recreated
   * elsewhere would otherwise poison every later call.
   */
  private async withMachine<T>(fn: (machine: Machine) => Promise<T>): Promise<T> {
    const machine = this.machine ?? (await this.client.resolve(this.name))
    this.machine = machine
    try {
      return await fn(machine)
    } catch (err) {
      if (!(err instanceof NotFoundError)) throw err
      this.client.forget(this.name)
      this.machine = undefined
      const fresh = await this.client.resolve(this.name)
      this.machine = fresh
      return await fn(fresh)
    }
  }
}

/** Per-argument POSIX quoting, so the quoted form re-yields the original. */
export function shellQuote(args: string[]): string {
  return args.map((a) => `'${a.replace(/'/g, "'\\''")}'`).join(' ')
}

function ndjsonResponse(body: unknown): Response {
  return new Response(JSON.stringify(body) + '\n', {
    status: 200,
    headers: { 'content-type': 'application/x-ndjson' },
  })
}

File: sdks/python/src/pilots/sprites_compat.py
"""``pilots.sprites_compat`` -- the sprites-shaped face of the pilots API.

It exists so a codebase written against ``sprites-py`` moves providers by
changing one import line. The surface is what a sprites client consumes and
nothing more: no services, no ``/control`` multiplex, no network policy.

Four rules decide the shapes here.

* **A sprite's ``id`` is the machine's NAME.** A sprites consumer persists the
  id and hands it straight back as a path segment. ``machine_id`` carries the
  ``m-…`` id for calls made through the typed client. Either form works as an
  argument: a value that matches no name but looks like a machine id is looked
  up as one.
* **``restore_checkpoint`` restores IN PLACE.** Exactly one request, and no
  machine is created. A machine created in a restore would get a new URL, and
  a URL is permanent.
* **``run`` returns a ``subprocess.CompletedProcess``**, which is what
  ``sprites-py`` returns and what a caller destructures.
* **``set_public_url`` is a no-op.** A workload's URL is public here by
  default, so there is nothing to switch on.
"""

from __future__ import annotations

import os
import subprocess
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from ._http import DEFAULT_TIMEOUT
from .client import PilotsClient
from .errors import NotFoundError, PilotsError
from .stream import ExecStream
from .types import Machine

__all__ = [
    "SpritesClient",
    "Sprite",
    "SpriteCommand",
    "SpriteFilesystem",
    "Checkpoint",
    "ExecResult",
    "URLSettings",
]


@dataclass
class URLSettings:
    """Accepted for source compatibility. ``auth`` is ``"public"`` or
    ``"sprite"``; pilots maps the latter onto ``url_auth="org"``."""

    auth: str = "public"
    private_access: str | None = None


@dataclass
class ExecResult:
    """What a sprites consumer destructures from an exec."""

    stdout: str
    stderr: str
    exit_code: int


@dataclass
class Checkpoint:
    id: str
    create_time: datetime | None = None
    comment: str = ""


def _looks_like_machine_id(ref: str) -> bool:
    return ref.startswith("m-") or ref.startswith("m_")


class SpriteFilesystem:
    """The sprites filesystem object, over exec. Reads and writes run as
    commands inside the machine (base64, so quotes, newlines and unicode
    survive), which keeps them consistent with everything else the agent does
    and correctly reverted by a checkpoint restore."""

    def __init__(self, sprite: Sprite, working_dir: str = "/") -> None:
        self._sprite = sprite
        self.working_dir = working_dir or "/"

    def _abs(self, path: str) -> str:
        if path.startswith("/"):
            return path
        base = self.working_dir.rstrip("/")
        return f"{base}/{path}" if base else f"/{path}"

    def read_text(self, path: str) -> str:
        import base64

        res = self._sprite._exec(f"base64 -w0 -- {_quote(self._abs(path))}")
        if res.exit_code != 0:
            raise PilotsError(res.stderr.strip() or f"cannot read {path}", status=0, next="check the path")
        return base64.b64decode(res.stdout).decode()

    def write_text(self, path: str, text: str) -> None:
        import base64

        target = self._abs(path)
        encoded = base64.b64encode(text.encode()).decode()
        res = self._sprite._exec(
            f"mkdir -p -- $(dirname {_quote(target)}) && printf %s {_quote(encoded)} | base64 -d > {_quote(target)}"
        )
        if res.exit_code != 0:
            raise PilotsError(res.stderr.strip() or f"cannot write {path}")

    def listdir(self, path: str = ".") -> list[str]:
        res = self._sprite._exec(f"ls -1A -- {_quote(self._abs(path))}")
        if res.exit_code != 0:
            raise PilotsError(res.stderr.strip() or f"cannot list {path}")
        return [line for line in res.stdout.splitlines() if line]

    def exists(self, path: str) -> bool:
        return self._sprite._exec(f"test -e {_quote(self._abs(path))}").exit_code == 0

    def remove(self, path: str) -> None:
        self._sprite._exec(f"rm -rf -- {_quote(self._abs(path))}")


def _quote(s: str) -> str:
    return "'" + s.replace("'", "'\\''") + "'"


class SpriteCommand:
    """The Go-style ``exec.Cmd`` face: build it, then take its output."""

    def __init__(self, sprite: Sprite, argv: list[str], *, cwd: str | None, env: dict[str, str] | None, tty: bool,
                 tty_rows: int | None, tty_cols: int | None, timeout: float | None) -> None:
        self._sprite = sprite
        self.argv = argv
        self.cwd = cwd
        self.env = env
        self.tty = tty
        self.tty_rows = tty_rows
        self.tty_cols = tty_cols
        self.timeout = timeout
        self.exit_code: int | None = None

    def _stream(self, stdin: bool = False) -> ExecStream:
        return self._sprite._client.machines.exec_stream(
            self._sprite.machine_id,
            self.argv,
            cwd=self.cwd,
            env=self.env,
            stdin=stdin or self.tty,
            tty=self.tty,
            rows=self.tty_rows,
            cols=self.tty_cols,
        )

    def output(self) -> bytes:
        """stdout. Raises when the command exited non-zero, as sprites does."""
        out, err, code = self._stream().output(self.timeout)
        self.exit_code = code
        if code != 0:
            raise PilotsError(err.decode(errors="replace").strip() or f"exit status {code}")
        return out

    def combined_output(self) -> bytes:
        out, err, code = self._stream().output(self.timeout)
        self.exit_code = code
        return out + err

    def run(self) -> int:
        _, _, code = self._stream().output(self.timeout)
        self.exit_code = code
        return code

    def start(self, stdin: bool = False) -> ExecStream:
        """The raw stream, for a caller that wants the pipes."""
        return self._stream(stdin=stdin)


class Sprite:
    """One machine, in sprites clothing."""

    def __init__(self, client: PilotsClient, machine: Machine | None, name: str) -> None:
        self._client = client
        self._machine = machine
        self._name = name

    # -- identity -------------------------------------------------------------

    @property
    def id(self) -> str:
        """The NAME. A sprites consumer persists this and hands it back."""
        return self._machine.name if self._machine else self._name

    @property
    def name(self) -> str:
        return self.id

    @property
    def machine_id(self) -> str:
        """The `m-…` id, for calls through the typed client."""
        return self._resolved().id

    @property
    def url(self) -> str:
        return self._resolved().url

    @property
    def status(self) -> str:
        return self._resolved().state

    @property
    def machine(self) -> Machine:
        return self._resolved()

    def _resolved(self) -> Machine:
        if self._machine is None:
            self._machine = _lookup(self._client, self._name)
        return self._machine

    def refresh(self) -> Sprite:
        self._machine = _lookup(self._client, self.id)
        return self

    # -- exec ------------------------------------------------------------------

    def _exec(self, cmd: str, *, cwd: str | None = None, env: dict[str, str] | None = None,
              timeout: float | None = None) -> ExecResult:
        res = self._client.machines.exec(
            self.machine_id, cmd=cmd, cwd=cwd, env=env, timeout_ms=int(timeout * 1000) if timeout else None
        )
        return ExecResult(res.stdout, res.stderr, res.exit_code)

    def run(
        self,
        *argv: str,
        capture_output: bool = False,
        timeout: float | None = None,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        check: bool = False,
    ) -> subprocess.CompletedProcess[bytes]:
        """``subprocess.run`` for a machine. Returns a ``CompletedProcess``
        whose ``stdout`` and ``stderr`` are bytes, as sprites-py does."""
        stream = self._client.machines.exec_stream(self.machine_id, list(argv), cwd=cwd, env=env)
        out, err, code = stream.output(timeout)
        if check and code != 0:
            raise subprocess.CalledProcessError(code, list(argv), out, err)
        return subprocess.CompletedProcess(
            list(argv), code, out if capture_output else None, err if capture_output else None
        )

    def command(
        self,
        *argv: str,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        tty: bool = False,
        tty_rows: int | None = None,
        tty_cols: int | None = None,
        timeout: float | None = None,
    ) -> SpriteCommand:
        return SpriteCommand(
            self, list(argv), cwd=cwd, env=env, tty=tty, tty_rows=tty_rows, tty_cols=tty_cols, timeout=timeout
        )

    def exec(self, cmd: str, **kw: Any) -> ExecResult:
        """The buffered form, taking a command line rather than an argv."""
        return self._exec(cmd, **kw)

    # -- filesystem ------------------------------------------------------------

    def filesystem(self, working_dir: str = "/") -> SpriteFilesystem:
        return SpriteFilesystem(self, working_dir)

    # -- checkpoints -----------------------------------------------------------

    def create_checkpoint(self, comment: str = "") -> Checkpoint:
        ck = self._client.machines.checkpoint(self.machine_id, comment or None)
        return Checkpoint(ck.id, _when(ck.created_at), ck.comment or "")

    def list_checkpoints(self) -> list[Checkpoint]:
        return [
            Checkpoint(c.id, _when(c.created_at), c.comment or "")
            for c in self._client.machines.list_checkpoints(self.machine_id)
        ]

    def restore_checkpoint(self, checkpoint_id: str) -> Sprite:
        """IN PLACE: the same machine, same id, same URL, same token."""
        self._machine = self._client.checkpoints.restore(checkpoint_id)
        return self

    # -- lifecycle -------------------------------------------------------------

    def suspend(self) -> None:
        self._client.machines.suspend(self.machine_id)

    def wake(self) -> None:
        self._client.machines.wake(self.machine_id)

    def destroy(self) -> None:
        self._client.machines.destroy(self.machine_id)

    delete = destroy

    def set_public_url(self, *_: Any, **__: Any) -> None:
        """A no-op: a machine's URL is public here by default."""
        return None

    def update_url_settings(self, settings: URLSettings) -> None:
        self._machine = self._client.machines.update(
            self.machine_id, {"url_auth": "org" if settings.auth == "sprite" else "public"}
        )


class SpritesClient:
    """The sprites-shaped client. ``token`` falls back to ``PILOT_API_KEY``
    (and ``SPRITES_TOKEN``, so an unedited environment still works)."""

    def __init__(
        self,
        token: str | None = None,
        *,
        base_url: str | None = None,
        timeout: float = DEFAULT_TIMEOUT,
        org: str | None = None,
        **_ignored: Any,
    ) -> None:
        key = token or os.environ.get("PILOT_API_KEY") or os.environ.get("SPRITES_TOKEN", "")
        self.client = PilotsClient(key, base_url=base_url, timeout=timeout, org=org)

    def close(self) -> None:
        self.client.close()

    def __enter__(self) -> SpritesClient:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    def sprite(self, name: str) -> Sprite:
        """A lazy handle: nothing is fetched until something needs the id."""
        return Sprite(self.client, None, name)

    def get_sprite(self, name: str) -> Sprite:
        return Sprite(self.client, _lookup(self.client, name), name)

    def create_sprite(
        self,
        name: str,
        *,
        url_settings: URLSettings | None = None,
        labels: list[str] | dict[str, str] | None = None,
        env: dict[str, str] | None = None,
        **_ignored: Any,
    ) -> Sprite:
        body: dict[str, Any] = {"name": name}
        if url_settings is not None and url_settings.auth == "sprite":
            body["url_auth"] = "org"
        if labels:
            # sprites takes a list of strings; pilots takes a map. A bare
            # string becomes a key with an empty value, which is what a tag is.
            body["labels"] = labels if isinstance(labels, dict) else {label: "" for label in labels}
        if env:
            body["env"] = env
        machine = self.client.machines.create(body)
        return Sprite(self.client, machine, machine.name)

    def list_sprites(self, prefix: str | None = None, **_ignored: Any) -> list[Sprite]:
        out = []
        for m in self.client.machines.list():
            if prefix and not m.name.startswith(prefix):
                continue
            out.append(Sprite(self.client, m, m.name))
        return out

    def destroy_sprite(self, name: str) -> None:
        self.sprite(name).destroy()

    delete_sprite = destroy_sprite


def _lookup(client: PilotsClient, ref: str) -> Machine:
    """A name, or a machine id. The alias serving sprites paths resolves
    names, so a name is tried first and an id is the fallback."""
    for m in client.machines.list():
        if m.name == ref:
            return m
    if _looks_like_machine_id(ref):
        return client.machines.get(ref)
    raise NotFoundError(f"no sprite {ref}", next="list_sprites shows what this key can see")


def _when(unix: int) -> datetime | None:
    return datetime.fromtimestamp(unix, tz=timezone.utc) if unix else None
